### PR TITLE
[SPARK-50301][SS] Make TransformWithState metrics reflect their intuitive meanings

### PR DESCRIPTION
**Pull request by neilramaswamy**
### What changes were proposed in this pull request?

These changes make the following changes to metrics in TWS:

- `allUpdatesTimeMs` now captures the time it takes to process all the new data with the user's stateful processor.
- `timerProcessingTimeMs` was added to capture the time it takes to process all the user's timers.
- `allRemovalsTimeMs` now captures the time it takes to do TTL cleanup at the end of a micro-batch.
- `commitTimeMs` now captures _only_ the time it takes to commit the state, not the TTL cleanup.

With these metrics, a user can have a fairly clear picture of where time is being spent in a micro-batch that uses TWS:

image

### Why are the changes needed?

The metrics today misrepresent what they're actually measuring.

### Does this PR introduce _any_ user-facing change?

Yes. Metrics for TWS are changing. However, since TWS is `private[sql]`, this shouldn't impact any real users.

### How was this patch tested?

We don't have any way to test these metrics in _any_ stateful operator for streaming today.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 1]
```diff
@@ -416,6 +429,8 @@ case class TransformWithStateExec(
       StatefulOperatorCustomSumMetric("numMapStateVars", "Number of map state variables"),
       StatefulOperatorCustomSumMetric("numDeletedStateVars", "Number of deleted state variables"),
       // metrics around timers
+      StatefulOperatorCustomSumMetric("timerProcessingTimeMs",
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:432]: Does this show up correctly in the progress metrics ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 2]
```diff
@@ -416,6 +429,8 @@ case class TransformWithStateExec(
...
+        "Number of milliseconds taken to process all timers"),
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:433]: Could we add some test to verify presence of metric/non-zero value (maybe add to an existing test) ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 3]
```diff
@@ -343,11 +343,20 @@ case class TransformWithStateExec(
     CompletionIterator[InternalRow, Iterator[InternalRow]] = {
     val allUpdatesTimeMs = longMetric("allUpdatesTimeMs")
     val commitTimeMs = longMetric("commitTimeMs")
-    val timeoutLatencyMs = longMetric("allRemovalsTimeMs")
+    val timerProcessingTimeMs = longMetric("timerProcessingTimeMs")
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:346]: Should we add another metric around the initial state rows ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 4]
```diff
@@ -343,11 +343,20 @@ case class TransformWithStateExec(
...
+    // allRemovalsTimeMs is not granular enough to distinguish between user-caused removals and
+    // TTL-caused removals. We could leave this empty and have two custom metrics, but leaving
+    // this as always 0 will be confusing for users. We could also time every call to clear(), but
+    // that could have performance penalties. So, we choose to capture TTL-only removals.
+    val allRemovalsTimeMs = longMetric("allRemovalsTimeMs")
 
     val currentTimeNs = System.nanoTime
     val updatesStartTimeNs = currentTimeNs
-    var timeoutProcessingStartTimeNs = currentTimeNs
+    var timerProcessingStartTimeNs = currentTimeNs
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:359]: Just to confirm - even the init state variants of the API will also record this as expected right ?

Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:359]: Correct. This method, `processDataWithPartition` is called _after_ `processInitialStateRows` is called. See the `processDataWithInitialState` method for this code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala (hunk 3 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:346]: ok sg

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala (hunk 2 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:433]: Yea sounds good - lets add this test

**Review by anishshri-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 5]
```diff
@@ -343,11 +343,20 @@ case class TransformWithStateExec(
     CompletionIterator[InternalRow, Iterator[InternalRow]] = {
     val allUpdatesTimeMs = longMetric("allUpdatesTimeMs")
     val commitTimeMs = longMetric("commitTimeMs")
-    val timeoutLatencyMs = longMetric("allRemovalsTimeMs")
+    val timerProcessingTimeMs = longMetric("timerProcessingTimeMs")
+    // In TWS, allRemovalsTimeMs is the time taken to remove state due to TTL.
+    // It does not measure any time taken by explicit calls from the user's state processor
+    // that clear()s state variables.
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:349]:
```suggestion
    // that clears state variables.
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 6]
```diff
@@ -387,13 +401,12 @@ case class TransformWithStateExec(
...
-      // Note: Due to the iterator lazy execution, this metric also captures the time taken
-      // by the upstream (consumer) operators in addition to the processing in this operator.
-      allUpdatesTimeMs += NANOSECONDS.toMillis(System.nanoTime - updatesStartTimeNs)
+      allRemovalsTimeMs += timeTakenMs {
+        processorHandle.doTtlCleanup()
+      }
+
       commitTimeMs += timeTakenMs {
         if (isStreaming) {
-          // clean up any expired user state
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:395]: Nice finding!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 7]
```diff
@@ -708,6 +732,52 @@ class TransformWithStateSuite extends StateStoreMetricsTest
...
+        assert(metrics.get("timerProcessingTimeMs") < 2000)
+      },
+
+      AddData(inputData, "b"),
+      AdvanceManualClock(1 * 1000),
+      // Side effect: timer scheduled for t = 2 + 10 = 12.
+      CheckNewAnswer(),
+      Execute { q =>
+        val metrics = q.lastProgress.stateOperators(0).customMetrics
+        assert(metrics.get("numRegisteredTimers") === 1)
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:761]: nit: it's not total but registered timers in this batch, right? It's "slightly" confusing, but isn't a scope of this PR to revisit.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala (hunk 5 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:349]: I missed this, lol Let's leave it as it is.

**Pull request closed without merging**